stat/lda.go

```diff
@@ -0,0 +1,208 @@
+// Package stat provides generalized statistical functions
+package stat
+
+import (
+	"fmt"
+	"math"
+	"sort"
+
+	"gonum.org/v1/gonum/mat"
+)
+
+// LD is a type for computing and extracting the linear discriminant analysis of a
+// matrix. The results of the linear discriminant analysis are only valid
+// if the call to LinearDiscriminant was successful.
+type LD struct {
+	n, p  int //n = row, p = col
+	k     int
+	ct    []float64  //Constant term of discriminant function of each class
+	mu    *mat.Dense //Mean vectors of each class
+	svd   *mat.SVD
+	ok    bool
+	eigen mat.EigenSym //Eigen values of common variance matrix
+}
+
+// LinearDiscriminant performs a linear discriminant analysis on the
+// matrix of the input data which is represented as an n×p matrix x where each
+// row is an observation and each column is a variable
+//
+// LinearDiscriminant returns whether the analysis was successful
+//
+// @param x is the training samples
+// @param y is the training labels in [0,k)
+// where k is the number of classes
+// @retun ok returns if whether the analysis was successful
+func (ld *LD) LinearDiscriminant(x mat.Matrix, y []int) (ok bool) {
+	ld.n, ld.p = x.Dims()
+	fmt.Printf("This is the matrix: %v \n", x)
+	fmt.Printf("This is the array: %v \n", y)
+	fmt.Printf("x dims: %v, %v \n", ld.n, ld.p)
+	if y != nil && len(y) != ld.n {
+		panic("The sizes of X and Y don't match")
+	}
+	var labels []int
+	var found bool
+	//Find unique labels
+	for i := 0; i < len(y); i++ {
+		found = false
+		for j := 0; j < len(labels); j++ {
+			if y[i] == labels[j] {
+				found = true
+				break
+			}
+		}
+		if !found {
+			labels = append(labels, y[i])
+		}
+
+	}
```

Comment:
In general, if you're iterating through a slice or array in Go, you probably should use `range()` instead of the c or java standard way. See: https://gobyexample.com/range

Also, I think in this case I'd simplify the code by using a `map` of type `bool` to keep track of which labels we've already encountered. This works because in Go when you reference an item that doesn't exist in a `map`, you get the "zero value" of the map's type (ie, in this case the zero value of `bool` is `false`). See below:

```Go
var labels []int
var labelMap = map[int]bool{}
for _, label := range(y) {
  if !labelMap[label] {
    labelMap[label] = true
    labels = append(labels, label)
  }
}
```